File: <100/008/008.cpp
#include<iostream>
#include<fstream>
using namespace std;
using ll = int64_t;

ll N, num_digits = 13;
string S, line, filename = "008.txt";

ll solve(){
  ll Max = 0;

  for(int i = 0; i < N-num_digits; ++i){
    ll m = 1;
    for(int d = 0; d < num_digits; ++d) m *= S[i+d] - '0';
    Max = max(Max, m);
  }

  return Max;
}

int main(){
  ifstream file(filename);

  while( getline(file, line) ){
    S += line;
  }

  N = S.size();
  cout << solve() << endl;
  return 0;
}

File: <100/008/008.hs
import System.IO ( hGetContents, openFile, IOMode(ReadMode) )
import Data.Char ( digitToInt, isDigit )


main :: IO ()
main = do
  handle   <- openFile filename ReadMode
  contents <- hGetContents handle
  let digitsString = extractDigits contents
  print $ solve digitsString


solve :: String -> Int
solve = solve' 0


solve' :: Int -> String -> Int
solve' n [] = n
solve' n s  = solve' m cs
  where
    m  = max n e
    e  = evalSeq s
    cs = tail s


-- "0b123\nx456\ncd789" -> "0123456789"
extractDigits :: String -> String
extractDigits [] = []
extractDigits (c:cs)
  | isDigit c = c : extractDigits cs
  | otherwise = extractDigits cs


-- evaluate the product of the string that consists of digits
evalSeq :: String -> Int
evalSeq s
  | length s < target = 0
  | otherwise         = product . map digitToInt $ take target s


----------------------- supplemental -----------------------
filename :: String
filename = "008.txt"


target :: Int
target   = 13
